multik-jvm/src/main/kotlin/org/jetbrains/kotlinx/multik/jvm/pludecomposition.kt

```diff
@@ -0,0 +1,259 @@
+package org.jetbrains.kotlinx.multik.jvm
+
+import org.jetbrains.kotlinx.multik.api.*
+import org.jetbrains.kotlinx.multik.ndarray.data.*
+import org.jetbrains.kotlinx.multik.ndarray.data.set
+import kotlin.math.abs
+import kotlin.math.min
+
+/**
+ * computes alpha * a * b + beta * c and stores result is c
+ * @param alpha scalar
+ * @param beta scalar
+ * @param a matrix
+ * @param b matrix
+ * @param c matrix
+ *
+ * blas / lapack: dgemm
+ */
+private fun dotThenPlusInplace(a: D2Array<Double>, b: D2Array<Double>, c: D2Array<Double>, alpha: Double, beta: Double) {
+    if(c.shape[0] <= 0 || c.shape[1] <= 0) return
+
+    if (alpha == 0.0) {
+        for (i in 0 until c.shape[0]) {
+            for (j in 0 until c.shape[1]) {
+                c[i, j] *= beta
+            }
+        }
+        return
+    }
+
+    for (j in 0 until c.shape[1]) {
+        for (i in 0 until c.shape[0]) {
+            c[i, j] *= beta
+        }
+
+        for (l in 0 until a.shape[1]) {
+            val temp = alpha * b[l, j]
+            for (i in 0 until c.shape[0]) {
+                c[i, j] += temp * a[i, l]
+            }
+        }
+    }
+}
+
+/**
+ * Solves ax=b equation where @param a lower triangular matrix with units on diagonal,
+ * rewrite @param b with solution
+ *
+ * notice: intentionally there is no checks that a[i, i] == 1.0 and a[i, >i] == 0.0,
+ * it is a contract of this method having no such checks
+ *
+ * lapack: dtrsm can do it (and have some extra options)
+ */
+private fun solveLowerTriangleSystemInplace(a: D2Array<Double>, b: D2Array<Double>) {
+    for (i in 0 until a.shape[1]) {
+        for (k in i+1 until a.shape[1]) {
+            for (j in 0 until b.shape[1]) {
+                // array getter have extra two bound checks, maybe better use b.data[...]
```

Comment:
create an issue for this problem

Reply:
I mean this checks are necessary for everyday usage. But to improve performance maybe better write directly to array memory